libr/core/windows_heap.c

```diff
@@ -0,0 +1,822 @@
+		// why it fails with 1000000 allocs? also with processes with segment heap enabled?
+		RtlDestroyQueryDebugBuffer (db);
+		r_sys_perror ("InitHeapInfo");
+		return NULL;
+	}
+	return db;
+}
+
+#define GROW_BLOCKS()										\
+	if (allocated <= count * sizeof (HeapBlockBasicInfo)) {	\
+		SIZE_T old_alloc = allocated;						\
+		allocated *= 2;										\
+		PVOID tmp = blocks;									\
+		blocks = realloc (blocks, allocated);				\
+		if (!blocks) {										\
+			free (tmp);										\
+			goto err;										\
+		}													\
+		memset ((BYTE *)blocks + old_alloc, 0, old_alloc);	\
+	}														\
+
+#define GROW_PBLOCKS()											\
+	if (*allocated <= *count * sizeof (HeapBlockBasicInfo)) {	\
+		SIZE_T old_alloc = *allocated;							\
+		*allocated *= 2;										\
+		PVOID tmp = *blocks;									\
+		tmp = realloc (*blocks, *allocated);					\
+		if (!tmp) {												\
+			return false;										\
+		}														\
+		*blocks = tmp;											\
+		memset ((BYTE *)(*blocks) + old_alloc, 0, old_alloc);	\
+	}															\
+
+static bool __lfh_segment_loop(HANDLE h_proc, PHeapBlockBasicInfo *blocks, SIZE_T *allocated, WPARAM lfhKey, WPARAM *count, WPARAM first, WPARAM next) {
+	while ((first != next) && next) {
+		HEAP_LFH_SUBSEGMENT subsegment;
+		ReadProcessMemory (h_proc, (void *)next, &subsegment, sizeof (HEAP_LFH_SUBSEGMENT), NULL);
+		subsegment.BlockOffsets.EncodedData ^= (DWORD)lfhKey ^ ((DWORD)next >> 0xC);
+		WPARAM mask = 1, offset = 0;
+		for (int l = 0; l < subsegment.BlockCount; l++) {
+			if (!mask) {
+				mask = 1;
+				offset++;
+				ReadProcessMemory (h_proc, (WPARAM *)(next + offsetof (HEAP_LFH_SUBSEGMENT, BlockBitmap)) + offset,
+					&subsegment.BlockBitmap, sizeof (WPARAM), NULL);
+			}
+			if (subsegment.BlockBitmap[0] & mask) {
+				GROW_PBLOCKS ();
+				WPARAM off = subsegment.BlockOffsets.FirstBlockOffset + l * subsegment.BlockOffsets.BlockSize;
+				(*blocks)[*count].address = next + off;
+				(*blocks)[*count].size = subsegment.BlockOffsets.BlockSize;
+				(*blocks)[*count].flags = 1 | SEGMENT_HEAP_BLOCK | LFH_BLOCK;
+				PHeapBlockExtraInfo extra = calloc (1, sizeof (HeapBlockExtraInfo));
+				extra->segment = next;
+				extra->granularity = sizeof (HEAP_ENTRY);
+				(*blocks)[*count].extra = EXTRA_FLAG | (WPARAM)extra;
+				*count += 1;
+			}
+			mask <<= 2;
+		}
+		next = (WPARAM)subsegment.ListEntry.Flink;
+	}
+	return true;
+}
+
+static bool GetSegmentHeapBlocks(HANDLE h_proc, PVOID heapBase, PHeapBlockBasicInfo *blocks, WPARAM *count, SIZE_T *allocated, WPARAM ntdllOffset) {
+	r_return_val_if_fail (h_proc && blocks && count && allocated, false);
+	WPARAM bytesRead;
+	SEGMENT_HEAP segheapHeader;
+	ReadProcessMemory (h_proc, heapBase, &segheapHeader, sizeof (SEGMENT_HEAP), &bytesRead);
+
+	if (segheapHeader.Signature != 0xddeeddee) {
+		return false;
+	}
+
+	WPARAM lfhKey;
+	WPARAM lfhKeyLocation = ntdllOffset + 0x44A8; // ntdll!RtlpHpHeapGlobals + 0x8
```

Comment:
magic number `0x44a8`? is this always constant? does `ntdll!RtlpHpHeapGlobals` ever change?

Reply:
at least on my system it is constant. I'm not confident in other windows versions though. And I don't know a way to get this offset at runtime without parsing a pdb file

Reply:
please make `RtlpHpHeapGlobals` offset into a variable so it's more clear